### PR TITLE
Fix for browser unfriendly characters in the nonce Url.

### PR DESCRIPTION
**Pull request by kumards**
At times, nonce generated using Convert.ToBase64String contains spaces. These type of nonces are altered by browser and ended up failing while decrypting nonce.

### CODE REVIEW

**Comment by Jetski5822** [comment]
is this fixing https://github.com/OrchardCMS/Orchard/issues/5111 ?

**Comment by kumards** [comment]
yeah, actually it was me who created that issue.

File: src/Orchard.Web/Modules/Orchard.Users/Services/UserService.cs [hunk 1]
```diff
@@ -85,7 +86,7 @@ ISiteService siteService
         public string CreateNonce(IUser user, TimeSpan delay) {
             var challengeToken = new XElement("n", new XAttribute("un", user.UserName), new XAttribute("utc", _clock.UtcNow.ToUniversalTime().Add(delay).ToString(CultureInfo.InvariantCulture))).ToString();
             var data = Encoding.UTF8.GetBytes(challengeToken);
-            return Convert.ToBase64String(_encryptionService.Encode(data));
+            return HttpServerUtility.UrlTokenEncode(_encryptionService.Encode(data));
```
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.Users/Services/UserService.cs]: I'd rather re-encode the base64 string. This way it's less breaking as currently working nonce won't be affected, and also still hides the content from plain sight.

**Comment by sebastienros** [comment]
Ping? And by the way can you target 1.9.x please?

File: src/Orchard.Web/Modules/Orchard.Users/Services/UserService.cs (hunk 1 above)
Comment (kumards) [inline src/Orchard.Web/Modules/Orchard.Users/Services/UserService.cs]: I don't understand what exactly you mean by "re-encode the base64 string".

Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.Users/Services/UserService.cs]: I meant `var data = _encryptionService.Decode(Convert.FromBase64String(HttpServerUtility.UrlTokenDecode(nonce)));` and `HttpServerUtility.UrlTokenEncode(Convert.ToBase64String(_encryptionService.Encode(data)))`

Comment (kumards) [inline src/Orchard.Web/Modules/Orchard.Users/Services/UserService.cs]: If I make the suggested code compile-able, I would end up with something like this.

``return 
HttpServerUtility.UrlTokenEncode(Convert.FromBase64String(Convert.ToBase64String(_encryptionService.Encode(data))));``
and
``var data = _encryptionService.Decode(Convert.FromBase64String(Convert.ToBase64String(HttpServerUtility.UrlTokenDecode(nonce))));``

That's repetitive conversion string to bytes and vice versa. Please bear with me, I am really missing something here.

**Comment by sebastienros** [comment]
Let's re-start the discussion.

I think the url encoding should be done in the controller, or wherever the nonce is used in a link. The `CreateNonce` and `DecryptNonce` will only manipulate a Base64 encoded value. The controller will `UrlEncode/Decode` it when necessary.

That should make things simpler to understand.

**Comment by kumards** [comment]
That makes lot of sense. Its the concern of the caller!

**Pull request closed without merging**